EvictionTest.testTTL_prolongationAfterNonTTLUpdate_Quick

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.map.EvictionTest.testTTL_prolongationAfterNonTTLUpdate_Quick(EvictionTest.java:144)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-OracleJDK1.6-freestyle/135/testReport/com.hazelcast.map/EvictionTest/testTTL_prolongationAfterNonTTLUpdate_Quick/

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-OracleJDK1.6-freestyle/134/testReport/junit/com.hazelcast.map/EvictionTest/testTTL_prolongationAfterNonTTLUpdate_Quick/

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.map.EvictionTest.testTTL_prolongationAfterNonTTLUpdate_Quick(EvictionTest.java:120)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-freestyle/1794/testReport/junit/com.hazelcast.map/EvictionTest/testTTL_prolongationAfterNonTTLUpdate_Quick/

```
Stacktrace
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.map.EvictionTest.testTTL_prolongationAfterNonTTLUpdate_Quick(EvictionTest.java:138)
Standard Output
Finished Running Test: testEvictionSpeedTestPerPartition in 10.402 seconds.
15:09:42,124  INFO |testTTL_prolongationAfterNonTTLUpdate_Quick| - [NodeExtension] Thread-62 - [127.0.0.1]:5062 [dev] [3.9.4-SNAPSHOT] Destroying node NodeExtension.
15:09:42,125  INFO |testTTL_prolongationAfterNonTTLUpdate_Quick| - [Node] Thread-62 - [127.0.0.1]:5062 [dev] [3.9.4-SNAPSHOT] Hazelcast Shutdown is completed in 121 ms.
15:09:42,125  INFO |testTTL_prolongationAfterNonTTLUpdate_Quick| - [LifecycleService] Thread-62 - [127.0.0.1]:5062 [dev] [3.9.4-SNAPSHOT] [127.0.0.1]:5062 is SHUTDOWN
BuildInfo right after testTTL_prolongationAfterNonTTLUpdate_Quick(com.hazelcast.map.EvictionTest): BuildInfo{version='3.9.4-SNAPSHOT', build='20180327', buildNumber=20180327, revision=b6615ab, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testTTL_prolongationAfterNonTTLUpdate_Quick(com.hazelcast.map.EvictionTest):'
15:09:35, accumulated pauses: 1094 ms, max pause: 194 ms, pauses over 1000 ms: 0
15:09:40, accumulated pauses: 1524 ms, max pause: 1396 ms, pauses over 1000 ms: 1

Started Running Test: testTTL_appliedFromLastUpdate
```

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/14735/testReport/junit/com.hazelcast.map/EvictionTest/testTTL_prolongationAfterNonTTLUpdate_Quick/

--- a/hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java
@@ -104,8 +104,9 @@ public class EvictionTest extends HazelcastTestSupport {
 
         map.put(1, "value0", 1, TimeUnit.SECONDS);
         map.put(1, "value1", 2, TimeUnit.SECONDS);
+        long sleepRef = System.currentTimeMillis();
         map.put(1, "value2", 300, TimeUnit.SECONDS);
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
 
         assertTrue(map.containsKey(1));
     }
@@ -114,15 +115,17 @@ public class EvictionTest extends HazelcastTestSupport {
     public void testTTL_prolongationAfterNonTTLUpdate_Quick() {
         final IMap<Integer, String> map = createSimpleMap();
 
+        long sleepRef = System.currentTimeMillis();
         map.put(1, "value0", 3, TimeUnit.SECONDS);
         // 1 second safety margin before eviction
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
 
+        sleepRef = System.currentTimeMillis();
         // this should prolong the life of the entry for another 3 seconds
         map.put(1, "value1");
         // 4 seconds of wait time in total, 1 second safety margin after a potential eviction
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
     }
 
@@ -131,27 +134,31 @@ public class EvictionTest extends HazelcastTestSupport {
     public void testTTL_prolongationAfterNonTTLUpdate_Slow() throws ExecutionException, InterruptedException {
         final IMap<Integer, String> map = createSimpleMap();
 
+        long sleepRef = System.currentTimeMillis();
         map.put(1, "value0", 3, TimeUnit.SECONDS);
         // 1 second safety margin before eviction
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
 
+        sleepRef = System.currentTimeMillis();
         // this should prolong the life of the entry for another 3 seconds
         map.put(1, "value1");
         // 4 seconds of wait time in total, 1 second safety margin after a potential eviction
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
 
+        sleepRef = System.currentTimeMillis();
         map.set(1, "value2");
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
 
         final HashMap<Integer, String> items = new HashMap<Integer, String>();
         items.put(1, "value3");
         items.put(2, "value1");
         items.put(3, "value1");
+        sleepRef = System.currentTimeMillis();
         map.putAll(items);
-        sleepSeconds(2);
+        sleepAtMostSeconds(sleepRef, 2);
         assertTrue(map.containsKey(1));
 
         map.putAsync(1, "value4").get();

--- a/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java
@@ -37,6 +37,8 @@ import com.hazelcast.internal.partition.InternalPartition;
 import com.hazelcast.internal.partition.InternalPartitionService;
 import com.hazelcast.internal.partition.impl.PartitionServiceState;
 import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
 import com.hazelcast.map.impl.operation.DefaultMapOperationProvider;
 import com.hazelcast.map.impl.operation.MapOperation;
 import com.hazelcast.map.impl.operation.MapOperationProvider;
@@ -360,6 +362,42 @@ public abstract class HazelcastTestSupport {
             TimeUnit.SECONDS.sleep(seconds);
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * Sleeps for time that is left until referenceTime + seconds. If no
+     * time is left until that time, a warning is logged and no sleep
+     * will happen.
+     * <p>
+     * Opposed to the language provided sleep constructs, this method
+     * does not guarantee minimum sleep time as it assumes that no time
+     * elapsed since {@code referenceTime} which is never true.
+     * <p>
+     * This method can be useful in hiding occasional hiccups of the
+     * execution environment in tests which are sensitive to
+     * oversleeping. Tests like the ones verify TTL, lease time behavior
+     * are typical examples for that.
+     *
+     * @param referenceTime the time in milliseconds since which the
+     *                      sleep end time should be calculated
+     * @param seconds       desired sleep duration in seconds
+     */
+    public static void sleepAtMostSeconds(long referenceTime, int seconds) {
+        long now = System.currentTimeMillis();
+        long sleepEnd = referenceTime + SECONDS.toMillis(seconds);
+        long sleepTime = sleepEnd - now;
+
+        if (sleepTime > 0) {
+            try {
+                MILLISECONDS.sleep(sleepTime);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        } else {
+            ILogger logger = Logger.getLogger(HazelcastTestSupport.class);
+            long absSleepTime = Math.abs(sleepTime);
+            logger.warning("There is no time left to sleep. We are beyond the desired end of sleep by " + absSleepTime + "ms");
         }
     }
 

